net/FlatBuffers/ByteBufferSlice.cs

```diff
@@ -0,0 +1,232 @@
+        public override unsafe void PutUlong(int offset, ulong value)
+        {
+            base.PutUlong(_off + offset, value);
+        }
+
+        public override unsafe void PutFloat(int offset, float value)
+        {
+            base.PutFloat(_off + offset, value);
+        }
+
+        public override unsafe void PutDouble(int offset, double value)
+        {
+            base.PutDouble(_off + offset, value);
+        }
+#else // !UNSAFE_BYTEBUFFER
+        public override void PutShort(int offset, short value)
+        {
+            base.PutShort(_off + offset, value);
+        }
+
+        public override void PutUshort(int offset, ushort value)
+        {
+            base.PutUshort(_off + offset, value);
+        }
+
+        public override void PutInt(int offset, int value)
+        {
+            base.PutInt(_off + offset, value);
+        }
+
+        public override void PutUint(int offset, uint value)
+        {
+            base.PutUint(_off + offset, value);
+        }
+
+        public override void PutLong(int offset, long value)
+        {
+            base.PutLong(_off + offset, value);
+        }
+
+        public override void PutUlong(int offset, ulong value)
+        {
+            base.PutUlong(_off + offset, value);
+        }
+
+        public override void PutFloat(int offset, float value)
+        {
+            base.PutFloat(_off + offset, value);
+        }
+
+        public override void PutDouble(int offset, double value)
+        {
+            base.PutDouble(_off + offset, value);
+        }
+#endif // UNSAFE_BYTEBUFFER
+
+        public override sbyte GetSbyte(int index)
+        {
+            return base.GetSbyte(_off + index);
+        }
+
+        public override byte Get(int index)
+        {
+            return base.Get(_off + index);
+        }
+
+        public override string GetStringUTF8(int startPos, int len)
+        {
+            return base.GetStringUTF8(_off + startPos, len);
+        }
+
+#if UNSAFE_BYTEBUFFER
+        public override unsafe ushort GetUshort(int offset)
+        {
+            return base.GetUshort(_off + offset);
+        }
+
+        public override unsafe uint GetUint(int offset)
+        {
+            return base.GetUint(_off + offset);
+        }
+
+        public override unsafe ulong GetUlong(int offset)
+        {
+            return base.GetUlong(_off + offset);
+        }
+
+        public override unsafe float GetFloat(int offset)
+        {
+            return base.GetFloat(_off + offset);
+        }
+
+        public override unsafe double GetDouble(int offset)
+        {
+            return base.GetDouble(_off + offset);
+        }
+#else // !UNSAFE_BYTEBUFFER
+        public override short GetShort(int index)
+        {
+            return base.GetShort(_off + index);
+        }
+
+        public override ushort GetUshort(int index)
+        {
+            return base.GetUshort(_off + index);
+        }
+
+        public override int GetInt(int index)
```

Comment:
This function is called a LOT, so it potentially becoming slower by being virtual I am afraid is not acceptable.

Reply:
Then I frankly do not see another way to support slices, which imho is the cleanest way of representing size-prefixed buffers. Note that using `ByteBuffer` in Java has exactly the same design, even for non-slices. The last resort would be do create a new `ByteBuffer` with a separate underlying `byte[]` array and copy all the previous contents to it, minus the size prefix. That would be a one-time overhead for creating the slice, and about 100% memory overhead, unless the caller discards all references to the size-prefixed buffer, and lets the GC handle that.

Reply:
Just FYI, I have added a small benchmark:

```csharp
[FlatBuffersTestMethod]
public void BenchmarkFlatBuffer()
{
    Random rnd = new Random(0);
    ByteBuffer bb = new ByteBuffer(File.ReadAllBytes(@"Resources/monsterdata_test.mon"));
    // bb = bb.Slice(); // toggle for ByteBufferSlice performance
    int maxIndex = bb.Length - 4;
    int sum = 0;

    Stopwatch sw = Stopwatch.StartNew();
    for (int i = 0; i < 1000000; ++i)
    {
        sum += bb.GetInt(rnd.Next(maxIndex));
    }
    sw.Stop();

    // write sum to avoid loop optimization
    Console.WriteLine("Regular: {0}, {1} GetInt ops/s", sum, 1000000.0 / sw.Elapsed.TotalMilliseconds);
}
```

I ran it on my Mac ten times with the regular `ByteBuffer` and with the `ByteBufferSlice`. Numbers are `GetInt` operations per second.

Version | Safe | Unsafe
--- | --- | ---
`ByteBuffer` | 33,017 | 44,017
`ByteBufferSlice` | 32,400 | 42,877

So there's a 1.9% slow-down for the safe implementation, and a 2.6% slow-down for the unsafe implementation.

Reply:
You say this is on your Mac.. does that mean these timings are from Mono, or the MS implementation?

If this is using the MS JIT, then indeed you have a good point that performance is acceptable for `_off`. Though part of this may be that the C# ByteBuffer is already quite slow.

I think we can do better though. If you look at how getting the root works in C# generated code:

`  public static Monster GetRootAsMonster(ByteBuffer _bb, Monster obj) { return (obj.__assign(_bb.GetInt(_bb.Position) + _bb.Position, _bb)); }`

This tells me, that to create a root from a size-prefixed buffer, all we need to do is pass it a `ByteBuffer` whose `_pos` has been moved forward by 4 bytes. All buffer accesses from there on are absolute, using the `bb_pos` stored in these table accessor structs as the starting point, so no need to keep adding `_off` at each access. And likely more robust since we can't forget to add `_off` somewhere.

Am I missing something?

Reply:
Hi, the timings are from Mono:
```bash
Mono JIT compiler version 5.4.1.7 (2017-06/e66d9abbb27 Wed Oct 25 12:10:41 EDT 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
```

Compiled with mcs:
```bash
Mono C# compiler version 5.4.1.0
```

I'm not sure there is an alternative on Mac? Should I get my hands on a Windows machine and benchmark it there as well?

I was thinking to separate `_pos` and `_off`, because `_pos` can be set externally, and therefore the size prefix could be restored (accidentally, e.g. by `Reset()`). For the use case where `bb_pos` is used as offset (when interacting with the `ByteBuffer` through `Monster`) you're correct, but since the `ByteBuffer` may be used rather freely I figured it'd be better to logically separate them.

I have added your suggestion, and these are the results:

Version | Safe | Unsafe
--- | --- | ---
original `ByteBuffer` (current `master`) | 33,876 | 44,436
`ByteBuffer` using `+ _off` | 33,790 (-0.3%) | 43,837 (-1.3%)
`ByteBuffer` using `virtual` | 33,017 (-2.5%) | 44,017 (-1.0%)
`ByteBufferSlice` using `override` | 32,400 (-4.4%) | 42,877 (-3.5%)
`ByteBuffer` using advanced `_pos` | 32,263 (-4.8%) | 45,593 (+2.6%)

I don't really know why this performs so poorly, I guess it's simply not the exact same system load today than it was last time. Your approach would certainly save a lot of code.

Reply:
Supposedly parts of MS's implementation are portable now and run on Linux, and I presume OS X as well, but haven't looked into it. It is entirely possible that Mono is not quite as optimized.

I'd prefer it if we use just a single variable, for efficency (possibly), especially since size-prefixing is such a minor feature compared to the base functionality. To screw this up, someone would have to load a size-prefixed buffer, slice it to obtain a new one, then call Reset on it, and somehow not understand that will get you your size prefixing back? That sounds unlikely to me.

I'm not sure if we're getting the correct results here. You're only calling `bb.GetInt` which doesn't even touch `_pos`! It directly uses its argument into `_buffer`.